pkg/util/provider/machinecontroller/machine_safety.go

```diff
@@ -32,29 +33,23 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
-	"k8s.io/klog/v2"
-)
-
-const (
-	// OverShootingReplicaCount freeze reason when replica count overshoots
-	OverShootingReplicaCount = "OverShootingReplicaCount"
-	// MachineDeploymentStateSync freeze reason when machineDeployment was found with inconsistent state
-	MachineDeploymentStateSync = "MachineDeploymentStateSync"
-	// TimeoutOccurred freeze reason when machineSet timeout occurs
-	TimeoutOccurred = "MachineSetTimeoutOccurred"
-	// UnfreezeAnnotation indicates the controllers to unfreeze this object
-	UnfreezeAnnotation = "safety.machine.sapcloud.io/unfreeze"
 )
 
 // reconcileClusterMachineSafetyOrphanVMs checks for any orphan VMs and deletes them
-func (c *controller) reconcileClusterMachineSafetyOrphanVMs(key string) error {
+func (c *controller) reconcileClusterMachineSafetyOrphanVMs(_ string) error {
 	ctx := context.Background()
 	reSyncAfter := c.safetyOptions.MachineSafetyOrphanVMsPeriod.Duration
 	defer c.machineSafetyOrphanVMsQueue.AddAfter("", reSyncAfter)
 
 	klog.V(3).Infof("reconcileClusterMachineSafetyOrphanVMs: Start")
 	defer klog.V(3).Infof("reconcileClusterMachineSafetyOrphanVMs: End, reSync-Period: %v", reSyncAfter)
 
+	// skip orphan collection in case of machine controller freeze
+	if c.safetyOptions.MachineControllerFrozen {
```

Comment:
can this check not be moved to the start of the method so that you exit early?
Do you need the following to be executed?
```go
reSyncAfter := c.safetyOptions.MachineSafetyOrphanVMsPeriod.Duration
defer c.machineSafetyOrphanVMsQueue.AddAfter("", reSyncAfter)
```

Reply:
yes I wanted it to be executed, as this is the default time after which we trigger orphan collection, and if its skipped, then orphan collection wouldn't be able to trigger on a periodic basis.

Reply:
As discussed there are 2 durations that are set in [options.go](https://github.com/gardener/machine-controller-manager/blob/0c341f69f4a0b548a9e827eb60955d3df4c93c56/cmd/machine-controller-manager/app/options/options.go#L76-L77):
```go
MachineSafetyAPIServerStatusCheckPeriod:  metav1.Duration{Duration: 1 * time.Minute},
MachineSafetyAPIServerStatusCheckTimeout: metav1.Duration{Duration: 30 * time.Second},
```
If i interpret this right, then API server connectivity will be checked every 1 min and each attempt will wait for a response from API server with a timeout of 30s. This check happens in a different go-routine than the code where the actual orphan collection happens. 

This means that in an edge case you can still have the possibility that within the total time of 1m30s while a go-routine waits for a response from a rather slow API server, another go-routine in MCM starts to collect orphans as it still does not see `frozen` state. So even with this change you will not be able to completely eliminate that situation but yes you can reduce the chances of it.

If this risk is acceptable then we can go with this simple change. If this risk is not acceptable then we need to discuss further.

Reply:
Yes you are right that in the worst case the marking of `frozen` condition can get delayed by 1m6sec and during that time if orphan collection happens, then nodes could get orphan collected.
I think this is a problem we can't avoid until we depend on polling approach and dealing with multiple goroutines.

Reply:
There is some timeframe between shutdown of the MCM in the source seed and the start up of the MCM in the destination seed since the etcd backups have to be copied after the kube-apiserver is shut down, and then there are a few other steps that take some time.
We observed that problem ~ 15-20 minutes after the migration had completed successfully so perhaps it's not so critical after all. 
Just one question - if MCM becomes frozen it will never unfreeze itself as long as the api-server constantly remains down right?

Reply:
Just to elaborate , you observed the problem after 15-20 minutes, because orphan collection runs 30minutes after the last orphan collection in worst case. In your case it happened to kick in 15-20 minutes after the migration. So in worst case it could happen just after the KAPI is down and the MCM in destination seed is not up, and in that 1min30sec all machines could get orphan collected. 

> Just one question - if MCM becomes frozen it will never unfreeze itself as long as the api-server constantly remains down right?

Yes it'll stay frozen until KAPI comes up